File: infra-pilot/services/orchestrator-agent/integration.py
"""Integration layer: database migrations, PostgreSQL helpers, and notification proxying."""

import asyncio
import logging
import os
from typing import Any, Dict, Optional

import requests
from config import config

logger = logging.getLogger(__name__)

REQUEST_TIMEOUT = 5


def _send_request(
    method: str, endpoint: str, data: Optional[Dict[str, Any]] = None
) -> tuple:
    """Send an HTTP request to the integration service.

    Args:
        method: HTTP method (GET or POST).
        endpoint: API endpoint path.
        data: Optional JSON payload for POST requests.

    Returns:
        A tuple of ``(success, response_data)``.
    """
    url = f"{config.INTEGRATION_SERVICE_URL}{endpoint}"
    try:
        if method == "GET":
            response = requests.get(url, timeout=REQUEST_TIMEOUT)
        elif method == "POST":
            response = requests.post(url, json=data, timeout=REQUEST_TIMEOUT)
        else:
            logger.warning("Unsupported HTTP method: %s", method)
            return False, None

        if response.status_code in (200, 201):
            return True, response.json()
        logger.warning(
            "Request failed with status %s: %s", response.status_code, response.text
        )
        return False, None
    except requests.Timeout:
        logger.warning("Request timeout for %s", endpoint)
        return False, None
    except requests.RequestException as exc:
        logger.warning("Request error for %s: %s", endpoint, exc)
        return False, None
    except ValueError as exc:
        logger.warning("Failed to parse response JSON: %s", exc)
        return False, None


def get_db_connection():
    """Create and return a new PostgreSQL database connection.

    Returns:
        A ``psycopg2.connection`` instance.
    """
    from db import get_sync_connection

    return get_sync_connection()


def _build_database_url() -> str:
    """Build a SQLAlchemy URL from DATABASE_URL or the DB_* config values."""
    database_url = os.getenv("DATABASE_URL", "")
    if database_url:
        return database_url.replace("postgres://", "postgresql://")
    return "postgresql://{user}:{password}@{host}:{port}/{name}".format(
        user=config.DB_USER,
        password=config.DB_PASSWORD,
        host=config.DB_HOST,
        port=config.DB_PORT,
        name=config.DB_NAME,
    )


def run_db_migrations() -> None:
    """Apply all pending Alembic migrations to the configured database.

    Migrations live in ``alembic/versions/`` and are the single source of
    truth for the schema. Run synchronously (blocking) at startup; the
    async wrapper below offloads it to a worker thread.
    """
    from alembic import command
    from alembic.config import Config

    service_root = os.path.dirname(os.path.abspath(__file__))
    cfg = Config(os.path.join(service_root, "alembic.ini"))
    cfg.set_main_option("script_location", os.path.join(service_root, "alembic"))
    cfg.set_main_option("sqlalchemy.url", _build_database_url())
    command.upgrade(cfg, "head")
    logger.info("Database migrations applied (upgrade head)")


async def init_database_tables():
    """Apply all pending database migrations.

    Kept as an async function with the historical name so existing call
    sites (main.py) stay unchanged. Runs ``alembic upgrade head``.
    """
    await asyncio.to_thread(run_db_migrations)


async def notify_integration(event_type: str, data: Dict[str, Any]) -> bool:
    """Send a server event notification to the integration service.

    Args:
        event_type: The type of event (e.g. ``server_created``).
        data: Event payload data.

    Returns:
        ``True`` if the notification was sent successfully.
    """
    payload = {
        "event_type": event_type,
        "server_name": data.get("server_name"),
        "details": data,
    }
    success, _ = _send_request("POST", "/api/notifications/server-event", payload)
    return success


async def notify_server_created(server_id: str, server_name: str) -> bool:
    """Notify that a server was created."""
    data = {
        "server_id": server_id,
        "server_name": server_name,
        "service": "orchestrator",
    }
    return await notify_integration("server_created", data)


async def notify_server_started(server_id: str, server_name: str) -> bool:
    """Notify that a server was started."""
    data = {
        "server_id": server_id,
        "server_name": server_name,
        "service": "orchestrator",
    }
    return await notify_integration("server_started", data)


async def notify_server_stopped(server_id: str, server_name: str) -> bool:
    """Notify that a server was stopped."""
    data = {
        "server_id": server_id,
        "server_name": server_name,
        "service": "orchestrator",
    }
    return await notify_integration("server_stopped", data)


async def notify_server_deleted(server_id: str, server_name: str) -> bool:
    """Notify that a server was deleted."""
    data = {
        "server_id": server_id,
        "server_name": server_name,
        "service": "orchestrator",
    }
    return await notify_integration("server_deleted", data)


async def sync_user_to_integration(
    user_id: str, email: str, username: str
) -> Dict[str, Any]:
    """Synchronise a user record to the integration service.

    Args:
        user_id: The Discord user ID.
        email: The user's email address.
        username: The user's display name.

    Returns:
        The API response dict, or an empty dict on failure.
    """
    payload = {"email": email, "username": username, "discord_id": user_id}
    success, response = _send_request("POST", "/api/users", payload)
    return response or {}


async def get_unified_metrics() -> Dict[str, Any]:
    """Fetch unified dashboard metrics from the integration service.

    Returns:
        The metrics response dict, or an empty dict on failure.
    """
    success, response = _send_request("GET", "/api/metrics/dashboard")
    return response or {}


async def broadcast_notification(message: str, title: str = "Notification") -> bool:
    """Broadcast a notification through the integration service.

    Args:
        message: The notification body.
        title: The notification title.

    Returns:
        ``True`` if broadcast was successful.
    """
    payload = {"content": message, "title": title}
    success, _ = _send_request("POST", "/api/notifications", payload)
    return success

File: infra-pilot/services/discord-service/integration.js
/**
 * @file Integration service client for the Discord bot.
 * Sends events and notifications to the external integration service.
 */

const axios = require('axios');

/** @constant {string} */
const INTEGRATION_SERVICE_URL = process.env.INTEGRATION_SERVICE_URL || 'http://localhost:9000';

/**
 * @typedef {Object} IntegrationEventData
 * @property {string} serverName
 * @property {string} [serverId]
 * @property {string} [error]
 */

/**
 * Send a server event notification to the integration service.
 * @param {string} eventType - Type of event (e.g. 'server_created')
 * @param {IntegrationEventData} data - Event payload
 * @returns {Promise<void>}
 */
async function notifyIntegration(eventType, data) {
  if (!eventType || !data) {
    console.error('[Integration] Missing eventType or data for notification');
    return;
  }

  try {
    await axios.post(`${INTEGRATION_SERVICE_URL}/api/notifications/server-event`, {
      event_type: eventType,
      server_name: data.serverName,
      details: data
    });
    console.log(`[Integration] Notified: ${eventType}`);
  } catch (error) {
    console.error(`[Integration] Failed to notify ${eventType}:`, error.message);
  }
}

/**
 * @param {string} serverName
 * @param {string} serverId
 * @returns {Promise<void>}
 */
async function notifyServerCreated(serverName, serverId) {
  return notifyIntegration('server_created', { serverName, serverId });
}

/**
 * @param {string} serverName
 * @param {string} serverId
 * @returns {Promise<void>}
 */
async function notifyServerStarted(serverName, serverId) {
  return notifyIntegration('server_started', { serverName, serverId });
}

/**
 * @param {string} serverName
 * @param {string} serverId
 * @returns {Promise<void>}
 */
async function notifyServerStopped(serverName, serverId) {
  return notifyIntegration('server_stopped', { serverName, serverId });
}

/**
 * @param {string} serverName
 * @param {string} serverId
 * @returns {Promise<void>}
 */
async function notifyServerDeleted(serverName, serverId) {
  return notifyIntegration('server_deleted', { serverName, serverId });
}

/**
 * @param {string} serverName
 * @param {string} serverId
 * @param {string} error
 * @returns {Promise<void>}
 */
async function notifyServerError(serverName, serverId, error) {
  return notifyIntegration('server_error', { serverName, serverId, error });
}

/**
 * Sync a user to the integration service.
 * @param {Object} userData - User data to sync
 * @returns {Promise<Object|null>} Synced user data or null on failure
 */
async function syncUserToIntegration(userData) {
  if (!userData || !userData.email) {
    console.error('[Integration] Missing userData or email for sync');
    return null;
  }

  try {
    const response = await axios.post(`${INTEGRATION_SERVICE_URL}/api/users`, userData);
    console.log(`[Integration] User synced: ${userData.email}`);
    return response.data;
  } catch (error) {
    console.error(`[Integration] User sync failed:`, error.message);
    return null;
  }
}

/**
 * Fetch unified dashboard metrics from the integration service.
 * @returns {Promise<Object|null>} Metrics data or null on failure
 */
async function getUnifiedMetrics() {
  try {
    const response = await axios.get(`${INTEGRATION_SERVICE_URL}/api/metrics/dashboard`);
    return response.data;
  } catch (error) {
    console.error(`[Integration] Metrics fetch failed:`, error.message);
    return null;
  }
}

/**
 * Broadcast a notification via the integration service.
 * @param {Object} message - Notification message payload
 * @returns {Promise<void>}
 */
async function broadcastNotification(message) {
  if (!message) {
    console.error('[Integration] Missing message for broadcast');
    return;
  }

  try {
    await axios.post(`${INTEGRATION_SERVICE_URL}/api/notifications`, message);
    console.log(`[Integration] Notification broadcast`);
  } catch (error) {
    console.error(`[Integration] Broadcast failed:`, error.message);
  }
}

/**
 * Send a Git push notification to a Discord channel via the integration service.
 * @param {string} repo - Repository name (e.g. 'owner/repo')
 * @param {string} branch - Branch name
 * @param {Array} [commits] - List of commits
 * @param {string} channelId - Discord channel ID
 * @returns {Promise<void>}
 */
async function notifyGitPush(repo, branch, commits, channelId) {
  if (!repo || !branch || !channelId) {
    console.error('[Integration] Missing required parameters for git push notification');
    return;
  }

  try {
    const commitLines = (commits || []).slice(0, 5).map(c =>
      `• \`${c.id?.slice(0, 7) || 'unknown'}\` ${c.message?.split('\n')[0] || 'no message'} — ${c.author?.name || 'unknown'}`
    ).join('\n');

    const message = {
      channel_id: channelId,
      embeds: [{
        title: `🚀 Push to ${repo}:${branch}`,
        description: commitLines || 'No commit details',
        color: 0x2ea043,
        timestamp: new Date().toISOString(),
        footer: { text: `${commits?.length || 0} commit(s) pushed` },
        url: `https://github.com/${repo}/commit/${commits?.[0]?.id || ''}`
      }]
    };

    await axios.post(`${INTEGRATION_SERVICE_URL}/api/notifications/discord`, message);
    console.log(`[Integration] Git push notification sent for ${repo}:${branch}`);
  } catch (error) {
    console.error(`[Integration] Git push notification failed for ${repo}:${branch}:`, error.message);
  }
}

module.exports = {
  notifyIntegration,
  notifyServerCreated,
  notifyServerStarted,
  notifyServerStopped,
  notifyServerDeleted,
  notifyServerError,
  syncUserToIntegration,
  getUnifiedMetrics,
  broadcastNotification,
  notifyGitPush
};